File: integration-test/java/jee/src/main/java/com/qualityminds/lazyval/integration/boundary/rest/LocaleRequestFilter.java
package com.qualityminds.lazyval.integration.boundary.rest;

import jakarta.ws.rs.container.ContainerRequestContext;
import jakarta.ws.rs.container.ContainerRequestFilter;
import jakarta.ws.rs.container.ContainerResponseContext;
import jakarta.ws.rs.container.ContainerResponseFilter;
import jakarta.ws.rs.ext.Provider;

import java.util.List;
import java.util.Locale;

/**
 * Resolves the request locale from the {@code Accept-Language} header and stores it in
 * {@link RequestLocaleHolder} so {@link RequestLocaleMessageInterpolator} can pick it up
 * when interpolating constraint violation messages.
 */
@Provider
public class LocaleRequestFilter implements ContainerRequestFilter, ContainerResponseFilter {

    @Override
    public void filter(ContainerRequestContext requestContext) {
        List<Locale> acceptable = requestContext.getAcceptableLanguages();
        Locale resolved = acceptable.stream()
                .filter(l -> !"*".equals(l.getLanguage()))
                .findFirst()
                .orElse(null);
        RequestLocaleHolder.set(resolved);
    }

    @Override
    public void filter(ContainerRequestContext requestContext, ContainerResponseContext responseContext) {
        RequestLocaleHolder.clear();
    }
}

File: integration-test/kotlin/jee/src/main/kotlin/com/qualityminds/lazyval/integration/boundary/rest/LocaleRequestFilter.kt
package com.qualityminds.lazyval.integration.boundary.rest

import jakarta.ws.rs.container.ContainerRequestContext
import jakarta.ws.rs.container.ContainerRequestFilter
import jakarta.ws.rs.container.ContainerResponseContext
import jakarta.ws.rs.container.ContainerResponseFilter
import jakarta.ws.rs.ext.Provider

/**
 * Resolves the request locale from the `Accept-Language` header and stores it in
 * [RequestLocaleHolder] so [RequestLocaleMessageInterpolator] can pick it up when
 * interpolating constraint violation messages.
 */
@Provider
class LocaleRequestFilter : ContainerRequestFilter, ContainerResponseFilter {

    override fun filter(requestContext: ContainerRequestContext) {
        val resolved = requestContext.acceptableLanguages
            .firstOrNull { it.language != "*" }
        RequestLocaleHolder.set(resolved)
    }

    override fun filter(
        requestContext: ContainerRequestContext,
        responseContext: ContainerResponseContext,
    ) {
        RequestLocaleHolder.clear()
    }
}
